keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+
+- https://github.com/kubernetes/kubernetes/issues/87857 (docs and understandability)
+- https://github.com/kubernetes/kubernetes/issues/87893 (holes in our test coverage matrix)
+- https://github.com/kubernetes/kubernetes/issues/85908 (failing tests, unclear semantics)
+- https://github.com/kubernetes/kubernetes/issues/86578 (needs e2e coverage)
+- https://github.com/kubernetes/kubernetes/issues/87709 (logging of netpol actions, will help describing states we reach) 
+- https://github.com/projectcalico/felix/issues/2032 (non-deterministic time frames for policy applications - addressable through published performance tests which measure time to policy implementation for several pods in several namespaces.)
+- https://github.com/projectcalico/felix/issues/2008 (need to test postStart pods in networkpolicy upstream.)
+- https://github.com/vmware-tanzu/antrea/issues/381 (addressable by performance tests *or* node-targeted tests.)
+- https://github.com/kubernetes/kubernetes/issues/88375 (the test matrix for Egress is almost entirely empty, decreasing the verbosity of new tests will organically increase likelihood of new test submissions over time.)
+
+
+### Consequences of this problem
+ 
+The consequences of this problem is that
+ 
+- CNI providers cannot easily be compared for functionality.
+- CNI providers implementing network policies must carry a lot of downstream test functionality.
+- Testing a CNI provider for Kubernetes compatibility requires a lot of interpretation and time investment.
+- Extending NetworkPolicy tests is time consuming and error prone, without a structured review process and acceptance standard.
+- It is hard to debug tests, due to the performance characteristics - pods are deleted after each test, so we cannot reproduce the state of the cluster easily.
+
+## A high level outline of Pod Traffic Pathways
+
+Before diving into test details, we outline different types of pod traffic.  Each one of these types of traffic may pose different bugs to a CNI provider.  
+
+Intranode
+- pod -> pod
+- pod -> host 
+- host -> pod
+
+Internode
+- pod -> pod
+- pod -> host
+- host -> pod
+- `*`host -> host (out-of-scope for K8s Network Policies API)
+- host -> pod (host networking)
+
+Traffic Transiting Service DNAT
+- Nodeport -> service (DNAT) -> pod
+- pod -> service (DNAT) -> pod
+
+Intranamespace
+- pod -> pod
+- pod (host networking) -> pod
+- pod -> pod (host networking)
+
+Internamespace
+- pod -> pod
+- pod (host networking) -> pod
+- pod -> pod (host networking)
+
+## Detailed examples of the problem statement
+ 
+### Incompleteness
```

Comment:
Not mentioned here, but another problem with incompleteness is that we don't generally test all of:

- pods/namespaces that exist before the policy were created obey the policy
- pods/namespaces that are created after the policy is created obey the policy
- pods/namespaces whose labels are changed such that they go from selected to not-selected or vice versa change their behavior appropriately.

(edit: also discussed later...)

Reply:
i added a small section on temporal testing here to, good point ,